products/cloudiot/api.yaml

```diff
@@ -192,3 +192,216 @@ objects:
                       output: true
                       description: |
                         The type of public key in the certificate.
+  - !ruby/object:Api::Resource
+    name: 'Device'
+    base_url: '{{registry}}/devices'
+    self_link: '{{registry}}/devices/{{name}}'
+    update_verb: :PATCH
+    update_mask: true
+    description: |
+      A Google Cloud IoT Core device.
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      guides:
+        'Official Documentation':
+          'https://cloud.google.com/iot/docs/'
+      api: 'https://cloud.google.com/iot/docs/reference/cloudiot/rest/'
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: registry
+        input: true
+        url_param_only: true
+        required: true
+        description: |
+          The name of the device registry where this device should be created.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'id'
+        input: true
+        required: true
+        description: |
+          The unique identifier for the device. For example,
+          `Device0`.
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          The resource path name. For example,
+          `projects/example-proj/locations/us-central1/registries/my-registry/devices/device0`.
+      - !ruby/object:Api::Type::String
+        name: 'numId'
+        output: true
+        description: |
+          A server-defined unique numeric ID for the device.
+          This is a more compact way to identify devices, and it is globally unique.
+      - !ruby/object:Api::Type::Array
+        name: 'credentials'
+        description: |
+          The credentials used to authenticate this device.
+        max_size: 3
+        item_type: !ruby/object:Api::Type::NestedObject
+          properties:
+            - !ruby/object:Api::Type::Time
+              name: 'expirationTime'
+              allow_empty_object: true
```

Comment:
I don't think we need `allow_empty_object` here. Is there a reason we would?

Reply:
It's optional in the [API](https://cloud.google.com/iot/docs/reference/cloudiot/rest/v1/projects.locations.registries.devices#devicecredential).

Should it be marked as `required: false` instead?

Reply:
Is it a field that a user can input? It is currently marked `output` which means we never allow a user to set the field. If a user can set it we should remove the `output` flag

`allow_empty_object` is a special case for a weird API behavior, it doesn't apply in this case. `required: false` is the same as not specifying `required` at all.

Reply:
I think it's an error! It's a parameter that the user can specify. So I have to remove both `allow_empty_object` and `output`, right? So that it's marked as an optional, input field, correct?

Reply:
Yes that should do it!

Reply:
ah! That's why I marked it as `optional` and `allow_empty_object`. If you don't specify any value for that field, the API returns `"expirationTime": "1970-01-01T00:00:00Z"` 🤦

Reply:
what should I do in this case? I think we've to keep the `output: true`

Reply:
(I removed `allow_empty_object: true`)

Reply:
Based on other examples (such as `ServiceAccountKey`'s ``validAfterTime` and `validBeforeTime`), I think that's what we want.

I verified that all the tests pass.

Reply:
`output: true` makes the field unsettable via terraform, meaning the API is the only one that can set a value for it. You can try to add the `expiration_time` field to a test and see what happens. 

I think what you want here is `default_from_api` instead of `output` which allows users to set the field, but also accepts an API-set value if the user did not specify one

Reply:
Thanks @slevenick ! Done. Removed `output: true` and added `default_from_api` in terraform overrieds. All tests pass!